### PR TITLE
Scrapper, Maria Kramkova - 22FPL1

### PR DESCRIPTION
**Pull request by kramkova**

### CODE REVIEW

**Comment by demid5111** [comment]
nothing to review

File: lab_5_scrapper/scrapper_config.json [hunk 1]
```diff
@@ -1,9 +1,13 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": ["https://www.fontanka.ru/24hours_news.html"],
+    "headers": {"user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 YaBrowser/24.1.0.0 Safari/537.36",
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json:3]: apply better formatting to JSON. for example, you can see how it should be formatted: https://jsonformatter.curiousconcept.com/#

File: lab_5_scrapper/scrapper.py [hunk 2]
```diff
@@ -204,12 +251,17 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    if not (os.path.isdir(ASSETS_PATH) and os.listdir(ASSETS_PATH)):
+        os.rmdir(ASSETS_PATH)
+    os.mkdir(ASSETS_PATH)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:256]: we use pathlib for filesystem operation, os module should not be used

File: lab_5_scrapper/scrapper_config.json [hunk 3]
```diff
@@ -1,9 +1,13 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": ["https://www.fontanka.ru/24hours_news.html"],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json:2]: one seed url is enough for 100 articles?

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -26,11 +38,36 @@ def _extract_config_content(self) -> ConfigDTO:
...
+                         should_verify_certificate=config['should_verify_certificate'],
+                         headless_mode=config['headless_mode']
+                         )
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters are not corrupt.
         """
+        config = self._extract_config_content()
+        if not all(seed.startswith(('https://www.', 'http://www.')) for seed in config.seed_urls):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:57]: we do not work with any website, right? we need seed urls to be exactly of the website we target

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -26,11 +38,36 @@ def _extract_config_content(self) -> ConfigDTO:
         Returns:
             ConfigDTO: Config values
         """
+        with open(self.path_to_config) as config_file:
+            config = json.load(config_file)
+        return ConfigDTO(seed_urls=config['seed_urls'],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:43]: rework to use kwargs-like style

File: lab_5_scrapper/scrapper.py [hunk 6]
```diff
@@ -26,11 +38,36 @@ def _extract_config_content(self) -> ConfigDTO:
...
+                         timeout=config['timeout'],
+                         should_verify_certificate=config['should_verify_certificate'],
+                         headless_mode=config['headless_mode']
+                         )
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters are not corrupt.
         """
+        config = self._extract_config_content()
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:56]: why do you do it twice? first DTO is created here and then again in init

**Review by demid5111 — CHANGES_REQUESTED** [review]

File: lab_5_scrapper/scrapper_config.json (hunk 3 above)
Comment (kramkova) [reply lab_5_scrapper/scrapper_config.json:2]: unfortunately, there is only one such page. I have added thematic sections, but another alternative is an archive (https://www.fontanka.ru/arc/news.html)

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (kramkova) [reply lab_5_scrapper/scrapper.py:57]: specified the mask

File: lab_5_scrapper/scrapper.py (hunk 6 above)
Comment (kramkova) [reply lab_5_scrapper/scrapper.py:56]: fixed

File: lab_5_scrapper/scrapper.py [hunk 7]
```diff
@@ -204,12 +244,17 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    if not ASSETS_PATH.exists():
+        shutil.rmtree(ASSETS_PATH)
+    ASSETS_PATH.mkdir(exist_ok=True, parents=True)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:249]: we just removed it, why do we need `exist_ok`?

**Review by demid5111 — CHANGES_REQUESTED** [review]
good start, but might be a bit late. Catch up with the remaining part of your group!

File: lab_5_scrapper/scrapper.py [hunk 8]
```diff
@@ -100,6 +136,9 @@ def make_request(url: str, config: Config) -> requests.models.Response:
     Returns:
         requests.models.Response: A response from a request
     """
+    headers = config.get_headers()
+    timeout = config.get_timeout()
+    return requests.get(url=url, headers=headers, timeout=timeout)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:141]: where are random waits between requests?

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -26,11 +38,28 @@ def _extract_config_content(self) -> ConfigDTO:
...
+        with open(self.path_to_config) as config_file:
+            config = json.load(config_file)
+        return ConfigDTO(**config)
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters are not corrupt.
         """
+        if not all(seed.startswith('https://www.fontanka.ru/') for seed in self.config.seed_urls):
+            raise Exception('IncorrectSeedURLError')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:50]: no. you need your own exceptions defined and then raised

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -39,6 +68,7 @@ def get_seed_urls(self) -> list[str]:
         Returns:
             list[str]: Seed urls
         """
+        return self.config.seed_urls
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:71]: some conceptual question. why do you think we create these getters that only return attributes of the aggregated object?

**Review by demid5111 — CHANGES_REQUESTED** [review]
for 4 looks like most of comments addressed. what is further?

**Comment by demid5111** [comment]
nothing to review

**Comment by demid5111** [comment]
nothing to review

**Comment by demid5111** [comment]
CI is red, you missed deadline. Now your penalty is -2. Do not worry, continue doing the task

**Review by demid5111 — APPROVED** [review]

File: .github/workflows/crawler.yml [hunk 11]
```diff
@@ -137,7 +137,7 @@ jobs:
       PR_NAME: ${{ github.event.pull_request.title }}
       PR_AUTHOR: ${{ github.actor }}
     runs-on: ubuntu-latest
-    timeout-minutes: 4
+    timeout-minutes: 10
```
Comment (demid5111) [inline .github/workflows/crawler.yml:140]: discard

File: requirements.txt [hunk 12]
```diff
@@ -1 +1,6 @@
-
+beautifulsoup4==4.12.3
+requests==2.31.0
+spacy-conll==3.4.0
+spacy-udpipe==1.0.0
+spacy==3.7.4
+stanza==1.8.2
```
Comment (demid5111) [inline requirements.txt:6]: it is not needed for lab 5, rm

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -195,6 +356,13 @@ def parse(self) -> Union[Article, bool, list]:
         Returns:
             Union[Article, bool, list]: Article instance
         """
+        response = make_request(url=self.full_url, config=self.config)
+        if not response.ok:
+            return False
+        soup = BeautifulSoup(response.text, features="lxml")
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:362]: list lxml in dependencies

File: lab_5_scrapper/scrapper.py [hunk 14]
```diff
@@ -127,11 +231,31 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+            if url not in self.urls:
+                break
+        return url
 
     def find_articles(self) -> None:
         """
         Find articles.
         """
+        goal = self.config.get_num_articles()
+        while len(self.urls) < goal:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:250]: are you sure about this condition? is it meaningful to make it a top-level condition?

File: lab_5_scrapper/scrapper.py [hunk 15]
```diff
@@ -127,11 +231,31 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+            for seed in self.get_search_urls():
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:251]: where do you check that obtained link is not yet present in a list?

File: .github/workflows/crawler.yml (hunk 11 above)
Comment (demid5111) [reply .github/workflows/crawler.yml:140]: CI takes so long just because you do not control number of collected URLs correctly

**Review by demid5111 — CHANGES_REQUESTED** [review]
CI is red. and when do you plan to start pipeline?

**Comment by demid5111** [comment]
nothing to review

**Pull request closed without merging**